Add xcm-simulator

evilrobot-01 opened the pull request:
Adds a `xcm-simulator`, based on the [example](https://github.com/paritytech/polkadot/tree/master/xcm/xcm-simulator/example) within the Polkadot repository, with Trappist-specific use cases as tests. Each closure scope in each test can be thought of as a block on the corresponding chain, with messages being dispatched to target chains via the mock message queue as it goes out of scope.

It effectively translates the integration tests in https://github.com/paritytech/trappist/pull/76 into Rust with the following main benefits:
- enables interactive debugging with breakpoints, allowing stepping through code to see exactly what is going on as messages are processed across chains
- uses existing runtime config as much as possible, simulating actual runtime interactions as best as possible
- shows xcm tracing output in test console for further additional insight

The tests can simply be run from the command line as follows, assuming one is in the root folder of the project, but the real value is running the tests with a debugger and breakpoints to explore how the instructions are handled by the `xcm-executor`. Note also that the crate is excluded from the workspace, effectively making it 'opt-in'.

```
cd xcm-simulator; cargo test; cd ..
```

Potential next steps:
- implement a helper function for determining fees, better demonstrating the relationships between xcm weights/fees
- add GH action to run when creating PRs, ensuring no regressions in cross-chain functionality

Any other suggestions for improvement welcome!

stiiifff reviewed (commented):
Great work @evilrobot-01, this is a super valuable addition to the repo.
Can you add instructions in the README on how to run the tests ?
Also, as you mentioned it, it would be nice to run these as part of a CI task.

evilrobot-01 commented:
Updated the PR description with a crude way to run the tests. Will update the readme and add a separate GH action for now, which can always be incorporated into the main CI flow as/when required.

bernardoaraujor commented on xcm-simulator/src/lib.rs:
```diff
@@ -0,0 +1,689 @@
+const A_PARA_ID: u32 = 1;
+decl_test_parachain! {
+	// Some generic parachain
+	pub struct ParaA {
```
what's the role of `ParaA`?

bernardoaraujor replied:
I'd give the variable a more insightful name, and also a more detailed description in the comments.

evilrobot-01 replied:
Renamed to `Template` to better describe intention, along with improved comments

stiiifff commented:
Fixes #103

bernardoaraujor commented on README.md:
````diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
+
+You can run all tests, including the tracing output, with:
+```
+cd xcm-simulator && cargo nextest run tests:: --no-capture --release
````
maybe add some contextual note about `cargo-nextest`?

or just instruct audience to use vanilla `cargo test`

evilrobot-01 replied:
Sorry, that was a copy-paste error and shouldn't have been pushed yet. Think it got caught up in the CI push I did to test that on GitHub. Will get it corrected to standard `cargo test`.

Thanks!

bernardoaraujor commented on README.md:
````diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
 ./zombienet-linux -p native spawn xcm-playground.toml
 ```
 
+### XCM Simulator
````
```suggestion
### Integration Tests
[parachains-integration-tests](https://github.com/paritytech/parachains-integration-tests) is a tool meant for XCM message execution in a locally spawned network. Tests are written as YAML files and converted into [Mocha](https://mochajs.org/) tests with [Chai](https://www.chaijs.com/) assertions.

The [`integration-tests`] directory has tests on Trappist use cases and instructions on how to run them.

### XCM Simulator
```

bernardoaraujor commented on README.md:
```diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
 
+### XCM Simulator
+The [XCM simulator](./xcm-simulator) can be used to further explore XCM message execution across the various runtimes used by Trappist. 
+Each Trappist use case is written as a test, allowing interactive debugging/exploration of message flows and instruction execution.
```
```suggestion
Each Trappist use case is written as a Rust unit test, allowing interactive debugging/exploration of message flows and instruction execution.
```

evilrobot-01 commented:
Updated the readme, thanks to additional information from @bernardoaraujor 🙌, and then also added a separate CI workflow which outputs the test results as an [artifact](https://github.com/paritytech/trappist/pull/102/checks?check_run_id=10560568569). I had hoped to also include the `xcm-executor` tracing output as further demonstration of what is available from the simulations, but `nextest` doesnt currently support this for passing tests so I have opened a feature request issue there.

PS - `nextest` used in CI as it allows converting results into junit format for nicer reporting using [test-reporter](https://github.com/dorny/test-reporter). Happy to hear of any better alternatives, haven't explored the GH test reporting landscape for a few months.

stiiifff commented:
@evilrobot-01 Super cool, I love the test report. Should we add a badge to the README with the latest "XCM tests" status ?

evilrobot-01 commented:
Sure, any preference on location?
